python/cuml/cuml/internals/base.pyx

```diff
@@ -471,6 +477,45 @@ class Base(TagsMixin,
                 func = nvtx_annotate(message=msg, domain="cuml_python")(func)
                 setattr(self, func_name, func)
 
+    @classmethod
+    def _hyperparam_translator(cls, **kwargs):
```

Comment:
I also think we could improve this a bit.

Before looking at how it was done in this PR my expectation was to find something like this:
```python
class Base:
  def _translate_hyperparameters(self, foo="some value", **hyper_parameters):
    if foo != "some value":
      hyper_parameters["foo"] = "some value"
    return hyper_parameters

class A(Base):
  def _translate_hyperparameters(self, bar=True, **hyper_parameters):
    hyper_parameters = super()._translate_hyperparameters(bar=bar, **hyper_parameters)
    # Update bar in case one of our parents modified it
    bar = hyper_parameters.get("bar", bar)
    if not bar:
      raise NotImplemented(f"Can't dispatch with '{bar=}', has to be True.")
    return hyper_parameters
```
So that we can accumulate translations from `Base` upwards, save on having the extra layer of the translation dictionary

(I need to think a bit more/try out the code, so bugs ahead but right now I'm channeling my best inner Raymond Hettinger: "there must be a better way" :D)

Reply:
Can we just use an `Enum` for `accept` and `dispatch`? To me it feels brittle if there's some estimator in the future that uses "accept" string as a valid hyperparam value.

Reply:
I'm not sure we have to worry about the case where the value is "accept" or "dispatch". The way I understand the dictionary like
```python
_hyperparam_interop_translator = {
        "solver": {
            "lbfgs": "qn",
            "liblinear": "qn",
            "newton-cg": "qn",
            "newton-cholesky": "qn",
            "sag": "qn",
            "saga": "qn"
        },
    }
```
is that it lists those hyper-parameters that needs translating and for those it lists the values and their translations. For example `solver="lbfgs"` needs translating to `solver="qn"`, but say `foobar=42` won't need translating because `foobar` isn't listed. Similarly `solver="dazzle"` doesn't need translating because it isn't listed. This makes me think we don't really need an entry with a value of `"accept"`, we could just have no entry (`"accept"` is the default assumption).

That is the `"accept"` magic value dealt with. Then there is the case of `"dispatch"` which is used for parameter values that can't be translated. It means "use scikit-learn". I think we should replace it by something like `NotImplemented` or some other exception or similar. This would deal with the case where a cuml value for a parameter should be `"dispatch"` - here we aren't able to tell if we should use scikit-learn or use cuml with this value. Hence lets use something like `NotImplemented`.

The parameter values used in scikit-learn don't matter because they are keys in the dictionaries.

Reply:
Having slept on [my comment](https://github.com/rapidsai/cuml/pull/6135/files/20c345bf43dc3a61a9923c8810b796e607b218e0#r1842237944) I've changed my mind. I still think having a method on each class that modifies the parameters as it sees fit could be nice. Not sure if it will create more typing/work than using a `dict` ¯\_(ツ)_/¯.


I also pondered the `_hyperparam_translator` function and I think this is how I'd write it. It doesn't use my suggestion from above (`NotImplemented`) but it could. The main changes are that it isn't a class method anymore (I couldn't work out why it was one), it merged the base classes translations with those of the derived class (I assume those are the only two interesting ones, we don't need to merge them together from the whole inheritance tree), I removed the `if` cases that were for `"accept"` (I think we don't do anything in those cases other than 👍 )

```python
    def _hyperparam_translator(self, **kwargs):
        """
        This method is meant to do checks and translations of hyperparameters
        at estimator creating time.
        Each children estimator can override the method, returning either
        modifier **kwargs with equivalent options, or
        """
        gpuaccel = True
        # Copy it so we can modify it
        translations = dict(super()._hyperparam_interop_translator)
        # Allow the derived class to overwrite the base class
        translations.update(self._hyperparam_interop_translator)
        for parameter_name, value in kwargs.items():
            # maybe clean up using: translations.get(parameter_name, {}).get(value, None)?
            if parameter_name in translations:
                if value in translations[parameter_name]:
                    if translations[parameter_name][value] == "dispatch":
                        gpuaccel = False
                    else:
                        kwargs[arg] = translations[parameter_name][value]

        return kwargs, gpuaccel
```

One more thought on `"dispatch"`: if we don't replace it with `NotImplemented` or similar, can we use `use_cpu` or something? I can't keep it straight in my head what "dispatch" means in the various libraries (most mean "use a GPU" when they talk about dispatching, in cuml it means "use a CPU"). Maybe something explicit like `"use_cpu"` makes it easier to reason about what is happening (thought my preference is still `NotImplemented` or similar).

Reply:
let's chat about this offline since it's getting long winded a bit, if my memory serves correctly, I got the idea of using the dict from you @betatim  😄